fastcrypto/src/tests/secp256r1_tests.rs

```diff
@@ -0,0 +1,446 @@
+        })
+        .unzip();
+
+    (digest.to_vec(), pubkeys, signatures)
+}
+
+#[test]
+fn verify_valid_batch() {
+    let (digest, pubkeys, signatures) = signature_test_inputs();
+
+    let res = Secp256r1PublicKey::verify_batch_empty_fail(&digest[..], &pubkeys, &signatures);
+    assert!(res.is_ok(), "{:?}", res);
+}
+
+#[test]
+fn verify_invalid_batch() {
+    let (digest, pubkeys, mut signatures) = signature_test_inputs();
+    // mangle one signature
+    signatures[0] = Secp256r1Signature::default();
+
+    let res = Secp256r1PublicKey::verify_batch_empty_fail(&digest, &pubkeys, &signatures);
+    assert!(res.is_err(), "{:?}", res);
+}
+
+#[test]
+fn verify_empty_batch() {
+    let (digest, _, _) = signature_test_inputs();
+
+    let res = Secp256r1PublicKey::verify_batch_empty_fail(&digest[..], &[], &[]);
+    assert!(res.is_err(), "{:?}", res);
+}
+
+#[test]
+fn verify_batch_missing_public_keys() {
+    let (digest, pubkeys, signatures) = signature_test_inputs();
+
+    // missing leading public keys
+    let res = Secp256r1PublicKey::verify_batch_empty_fail(&digest, &pubkeys[1..], &signatures);
+    assert!(res.is_err(), "{:?}", res);
+
+    // missing trailing public keys
+    let res = Secp256r1PublicKey::verify_batch_empty_fail(
+        &digest,
+        &pubkeys[..pubkeys.len() - 1],
+        &signatures,
+    );
+    assert!(res.is_err(), "{:?}", res);
+}
+
+#[test]
+fn verify_invalid_signature() {
+    // Get a keypair.
+    let kp = keys().pop().unwrap();
+
+    // Make signature.
+    let message: &[u8] = b"Hello, world!";
+    let digest = Sha256::digest(message);
+
+    // Verify the signature against good digest passes.
+    let signature = kp.sign(digest.as_ref());
+    assert!(kp.public().verify(digest.as_ref(), &signature).is_ok());
+
+    // Verify the signature against bad digest fails.
+    let bad_message: &[u8] = b"Bad message!";
+    let digest = Sha256::digest(bad_message);
+
+    assert!(kp.public().verify(digest.as_ref(), &signature).is_err());
+}
+
+#[test]
+fn verify_valid_batch_different_msg() {
+    let inputs = signature_tests::signature_test_inputs_different_msg::<Secp256r1KeyPair>();
+    let res = Secp256r1PublicKey::verify_batch_empty_fail_different_msg(
+        &inputs.digests,
+        &inputs.pubkeys,
+        &inputs.signatures,
+    );
+    assert!(res.is_ok(), "{:?}", res);
+}
+
+#[test]
+fn verify_invalid_batch_different_msg() {
+    let mut inputs = signature_tests::signature_test_inputs_different_msg::<Secp256r1KeyPair>();
+    inputs.signatures[0] = Secp256r1Signature::default();
+    let res = Secp256r1PublicKey::verify_batch_empty_fail_different_msg(
+        &inputs.digests,
+        &inputs.pubkeys,
+        &inputs.signatures,
+    );
+    assert!(res.is_err(), "{:?}", res);
+}
+
+// #[test]
+// fn fail_to_verify_if_upper_s() {
+//     // Test case from https://github.com/fjl/go-ethereum/blob/41c854a60fad2ad9bb732857445624c7214541db/crypto/signature_test.go#L79
```

Comment:
Same: please add commentary offering forward guidance on what we want this test to be in the future.

Reply:
The test is now fixed.